contracts/utils/InterchainAddressTracker.sol

```diff
@@ -0,0 +1,156 @@
+     */
+    function chainName() external view returns (string memory chainName_) {
+        chainName_ = StringStorage.get(_CHAIN_NAME_SLOT);
+    }
+
+    /**
+     * @dev Gets the key for the trusted address at a remote chain
+     * @param chain Chain name of the remote chain
+     * @return slot the slot to store the trusted address in
+     */
+    function _getTrustedAddressSlot(string memory chain) internal pure returns (bytes32 slot) {
+        slot = keccak256(abi.encode(PREFIX_ADDRESS_MAPPING, chain));
+    }
+
+    /**
+     * @dev Gets the key for the trusted address at a remote chain
+     * @param chain Chain name of the remote chain
+     * @return slot the slot to store the trusted address hash in
+     */
+    function _getTrustedAddressHashSlot(string memory chain) internal pure returns (bytes32 slot) {
+        slot = keccak256(abi.encode(PREFIX_ADDRESS_HASH_MAPPING, chain));
+    }
+
+    /**
+     * @dev Sets the trusted address and its hash for a remote chain
+     * @param chain Chain name of the remote chain
+     * @param trustedAddress the string representation of the trusted address
+     */
+    function _setTrustedAddress(string memory chain, string memory trustedAddress) internal {
+        StringStorage.set(_getTrustedAddressSlot(chain), trustedAddress);
+
+        bytes32 slot = _getTrustedAddressHashSlot(chain);
+        bytes32 addressHash = keccak256(bytes(trustedAddress));
+        assembly {
+            sstore(slot, addressHash)
+        }
+    }
+
+    /**
+     * @dev Gets the trusted address at a remote chain
+     * @param chain Chain name of the remote chain
+     * @return trustedAddress the trusted address for the chain. Returns '' if the chain is untrusted
+     */
+    function getTrustedAddress(string memory chain) public view returns (string memory trustedAddress) {
+        trustedAddress = StringStorage.get(_getTrustedAddressSlot(chain));
+    }
+
+    /**
+     * @dev Gets  the trusted address hash at a remote chain
+     * @param chain Chain name of the remote chain
+     * @return trustedAddressHash the hash of the trusted address
+     */
+    function getTrustedAddressHash(string memory chain) public view returns (bytes32 trustedAddressHash) {
+        bytes32 slot = _getTrustedAddressHashSlot(chain);
+        assembly {
+            trustedAddressHash := sload(slot)
+        }
+    }
+
+    /**
+     * @dev Validates that the sender is a valid interchain token service address
+     * @param sourceChain Source chain of the transaction
+     * @param sourceAddress Source address of the transaction
+     * @return bool true if the sender is validated, false otherwise
+     */
+    function validateSender(string calldata sourceChain, string calldata sourceAddress) external view returns (bool) {
+        bytes32 sourceAddressHash = keccak256(bytes(sourceAddress));
+
+        return sourceAddressHash == getTrustedAddressHash(sourceChain);
```

Comment:
Being a public method, will `getTrustedAddressHash` cause a delegatecall via a proxy? Should we compute it internally instead?

Reply:
public functions are just internal when called from within the contract, they can just also be called from outside.